### PR TITLE
chapter 11

### PR DESCRIPTION
**Pull request by chan227**

### CODE REVIEW

**Comment by chan227** [comment]
ready

File: app/helpers/account_activations_helper.rb [hunk 1]
```diff
@@ -0,0 +1,2 @@
+module AccountActivationsHelper
```
Comment (ngocvt-0484) [inline app/helpers/account_activations_helper.rb]: please delete this file

File: app/mailers/user_mailer.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+class UserMailer < ApplicationMailer
+
```
Comment (ngocvt-0484) [inline app/mailers/user_mailer.rb]: remove empty line

File: app/models/user.rb [hunk 3]
```diff
@@ -34,12 +36,33 @@ def remember
...
+    BCrypt::Password.new(digest).is_password? token
   end
 
   def forget
     update_attribute :remember_digest, nil
   end
+
+  def activate
+    update_attributes :activated, true
+    update_attributes :activated_at, Time.zone.now
```
Comment (ngocvt-0484) [inline app/models/user.rb]:
```suggestion
    update_attributes activated_at: Time.zone.now, activated: true
```
remove update_attributes :activated, true, we can update two attributes in same time by update_attributes
you can learn about other types of updates like: update_attribute, update_column, update_columns

File: app/controllers/account_activations_controller.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
+class AccountActivationsController < ApplicationController
+
```
Comment (ngocvt-0484) [inline app/controllers/account_activations_controller.rb]: remove this line

File: app/assets/stylesheets/account_activations.scss [hunk 5]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the AccountActivations controller here.
```
Comment (ngocvt-0484) [inline app/assets/stylesheets/account_activations.scss]: remove this file

File: app/helpers/account_activations_helper.rb (hunk 1 above)
Comment (chan227) [reply app/helpers/account_activations_helper.rb]: done

File: app/mailers/user_mailer.rb (hunk 2 above)
Comment (chan227) [reply app/mailers/user_mailer.rb]: done

File: app/models/user.rb (hunk 3 above)
Comment (chan227) [reply app/models/user.rb]: Noted it, thank you.

File: app/controllers/account_activations_controller.rb (hunk 4 above)
Comment (chan227) [reply app/controllers/account_activations_controller.rb]: done

File: app/assets/stylesheets/account_activations.scss (hunk 5 above)
Comment (chan227) [reply app/assets/stylesheets/account_activations.scss]: done

**Comment by chan227** [comment]
ready

**Comment by chan227** [comment]
ready

File: app/controllers/sessions_controller.rb [hunk 6]
```diff
@@ -4,9 +4,14 @@ def new; end
   def create
     user = User.find_by email: params[:session][:email].downcase
     if user&.authenticate params[:session][:password]
-      log_in user
-      params[:session][:remember_me] == "1" ? remember(user) : forget(user)
-      redirect_back_or user
+      if user.activated?
+        log_in user
+        params[:session][:remember_me] == '1' ? remember(user) : forget(user)
```
Comment (manhnt-0960) [inline app/controllers/sessions_controller.rb]: use `"` instead of `'`

File: app/models/user.rb [hunk 7]
```diff
@@ -34,12 +36,32 @@ def remember
     update_attribute :remember_digest, User.digest(remember_token)
   end
 
-  def authenticated? remember_token
-    return false unless remember_digest
-    BCrypt::Password.new(remember_digest).is_password?remember_token
+  def authenticated? attribute, token
+    digest = send "#{attribute}_digest"
+    return false if digest.nil?
```
Comment (manhnt-0960) [inline app/models/user.rb]:
```suggestion
    return false unless digest
```

File: app/controllers/sessions_controller.rb (hunk 6 above)
Comment (chan227) [reply app/controllers/sessions_controller.rb]: done

File: app/models/user.rb (hunk 7 above)
Comment (chan227) [reply app/models/user.rb]: done

**Comment by chan227** [comment]
ready

**Comment by manhnt-0960** [comment]
merged

**Pull request merged**